src/client_session.cpp

```diff
@@ -50,95 +50,79 @@ ClientSession::ClientSession(const string & db_name, unsigned long long foregrou
     this->m_current_session_id = "";
     this->m_session_index = 0;
   }
+  this->update_last_session_check_at();
 
   session_rows->clear();
   delete(session_rows);
-
-  this->update_session();
-  this->update_accessed_last();
-  this->update_session_context_data();
-  Storage::instance()->insert_update_session(this->m_session_context_data);
-}
-
-ClientSession::~ClientSession() {
-  this->stop();
 }
 
 // --- Public
 
-void ClientSession::start() {
-  lock_guard<mutex> guard(this->m_run_check);
-  if (this->m_is_running) {
-    return;
-  }
-  this->m_is_running = true;
-  this->m_daemon_thread = thread(&ClientSession::run, this);
-}
-
-void ClientSession::stop() {
-  unique_lock<mutex> locker(this->m_run_check);
-  if (this->m_is_running == true) {
-    this->m_is_running = false;
-    locker.unlock();
-    this->m_daemon_thread.join();
-  } else {
-    locker.unlock();
-  }
+void ClientSession::start_new_session() {
+  this->m_is_new_session = true;
 }
 
-SelfDescribingJson ClientSession::get_session_context(const string & event_id) {
-  lock_guard<mutex> guard(this->m_safe_get);
+SelfDescribingJson ClientSession::update_and_get_session_context(const string & event_id) {
+  json session_context_data;
+  bool save_to_storage = false;
 
-  this->update_accessed_last();
+  {
+    lock_guard<mutex> guard(this->m_safe_get);
 
-  if (this->m_first_event_id == "") {
-    this->m_first_event_id = event_id;
-    this->m_session_context_data[SNOWPLOW_SESSION_FIRST_ID] = this->m_first_event_id;
+    if (this->should_update_session()) {
+      this->update_session(event_id);
+      save_to_storage = true;
+    }
+    this->update_last_session_check_at();
+    session_context_data = this->m_session_context_data;
   }
 
-  SelfDescribingJson sdj(SNOWPLOW_SCHEMA_CLIENT_SESSION, this->m_session_context_data);
+  if (save_to_storage) {
+    Storage::instance()->insert_update_session(this->m_session_context_data);
+  }
```

Comment:
The `save_to_storage` is set inside the lock but used later. In case of multiple threads I think it would be possible (even if rare) that the `save_to_storage` is set to false again before that this piece of code has been executed.
Would it make sense to set it to false in this block?
Also, should we lock again in order to avoid conflicts with session_context_data? (e.g.: reading during a writing operation)

Reply:
We talked on Slack about this – the `save_to_storage` is a local variable so there isn't a multithreading problem.

However, your second point about reading while writing is valid, changed to save from a local variable instead of the instance one, thanks!